cli/cmd/cluster/certificate-rotator.go

```diff
@@ -0,0 +1,244 @@
+	// reconciliation period and time spent in reconciliation loop (based on number of workloads
+	// in the cluster).
+	kubeControllerManagerReconciliationPeriod = 20 * time.Second
+)
+
+type CertificateRotateOptions struct {
+	Confirm    bool
+	Verbose    bool
+	ConfigPath string
+	ValuesPath string
+}
+
+func RotateCertificates(contextLogger *log.Entry, options CertificateRotateOptions) error {
+	cc := clusterConfig{
+		verbose:    options.Verbose,
+		configPath: options.ConfigPath,
+		valuesPath: options.ValuesPath,
+	}
+
+	c, err := cc.initialize(contextLogger)
+	if err != nil {
+		return fmt.Errorf("initializing: %w", err)
+	}
+
+	exists, err := clusterExists(c.terraformExecutor)
+	if err != nil {
+		return fmt.Errorf("checking if cluster exists: %w", err)
+	}
+	if !exists {
+		return errors.New("cannot rotate certificates on a non-existent cluster")
+	}
+	if c.platform.Meta().Managed {
+		// TODO: do we want to error here?
+		return errors.New("cannot rotate certificates on a managed cluster")
+	}
+
+	// Tainting certificates so they get rotated
+	if err := c.taintCertificates(); err != nil {
+		return fmt.Errorf("tainting certificate resources: %w", err)
+	}
+
+	// Apply the terraform changes to replace tainted resources
+	if err := c.platform.Apply(&c.terraformExecutor); err != nil {
+		return fmt.Errorf("applying platform: %v", err)
+	}
+
+	kg := kubeconfigGetter{
+		platformRequired: true,
+		clusterConfig:    cc,
+	}
+	kubeconfig, err := kg.getKubeconfig(contextLogger, c.lokomotiveConfig)
+	if err != nil {
+		return fmt.Errorf("getting kubeconfig: %v", err)
+	}
+
+	fmt.Printf("\nApplying a controlplane update with the new CA.\n")
+	if err := c.upgradeControlPlane(contextLogger, kubeconfig); err != nil {
+		return fmt.Errorf("running controlplane upgrade: %v", err)
+	}
+
+	cs, err := k8sutil.NewClientset(kubeconfig)
+	if err != nil {
+		return fmt.Errorf("creating clientset from kubeconfig: %w", err)
+	}
+
+	newCACert, err := c.readKubernetesCAFromTerraformOutput()
+	if err != nil {
+		return fmt.Errorf("reading Kubernetes CA certificate from Terraform output: %w", err)
+	}
+
+	crConfig := certificateRotatorConfig{
+		clientSet:            cs,
+		newCACert:            newCACert,
+		logger:               contextLogger,
+		daemonSetsToRestart:  c.platform.Meta().DaemonSets,
+		deploymentsToRestart: c.platform.Meta().Deployments,
+	}
+
+	cr, err := newCertificateRotator(crConfig)
+	if err != nil {
+		return fmt.Errorf("preparing certificate rotator: %w", err)
+	}
+
+	if err := cr.rotate(); err != nil {
+		return fmt.Errorf("rotating certificates: %w", err)
+	}
+
+	return nil
+}
+
+func newCertificateRotator(config certificateRotatorConfig) (*certificateRotator, error) {
+	if config.clientSet == nil {
+		return nil, fmt.Errorf("clientSet can't be nil")
+	}
+
+	if config.newCACert == "" {
+		return nil, fmt.Errorf("new CA certificate can't be empty")
+	}
+
+	return &certificateRotator{
+		config: config,
+	}, nil
+}
```

Comment:
Given that we use unexported struct here, maybe we can just have a `certificateRotator` struct and a validate method on it, called with best effort?